Add basic warnings

brgr opened the pull request:
Generate Warnings for float and add a regression test where it warns if a float (double) value might be Nan or +/-infinity -- i.e. we have top

brgr committed:
add some implementation for the "norm" method in FloatDomain, however at the moment incorrect

Now, the norm method is called at eval_binop and successfully fails,
which currently is wanted like this, just to test that it works.
We need to change it now s.t. it prints a warning instead of an error.

brgr committed:
Rework norm slightly s.t. it uses set_overflow_flag now, which however is still incorrect

brgr committed:
Add FloatMessage message category in messageCategory.ml

Also add it in options.schema.json

brgr committed:
Add a todo note for later in sarifRules.ml

brgr committed:
Add a regression test for float that expects a warning for division by zeron / infinity

brgr committed:
Clean up floatDomain.ml slightly: Remove TODOs/FIXMEs and inline message warning

brgr committed:
Fix regression test naming and merge them together

brgr committed:
Remove FloatMessage module in messageCategory.ml, as it's not needed anymore

brgr committed:
Remove todo message

brgr committed:
Remove todo message

brgr committed:
Remove todo comments and/or clean up some stuff

Dudeldu commented on tests/regression/56-floats/03-infinity_or_nan.c:
```diff
@@ -0,0 +1,11 @@
+void main()
+{
+    double data;
+    fscanf(stdin, "%lf", &data);
```
i know, it was me coming up with that originally, however we are not checking for uninitialized-access. Therefore, we should probably leaf this out, and make minimal examples for our cases.

brgr replied:
Good point. Done.

Dudeldu commented on src/cdomains/valueDomain.ml:
```diff
@@ -193,7 +193,7 @@ struct
       match t with
       | TInt (ikind, _) -> `Int (ID.of_int ikind BI.zero)
-      | TFloat (FDouble, _) -> failwith "todo"
+      | TFloat (FDouble, _) -> failwith "todo zero_init_value"
```
Undo this unrelated change

brgr replied:
Done

Dudeldu commented on src/cdomains/floatDomain.ml:
```diff
@@ -37,18 +37,20 @@ module type FloatDomainBase = sig
+        else None
+      | _ -> None
+    in if normed == None then 
+      Messages.warn ~category:Messages.Category.FloatMessage ~tags:[CWE 189; CWE 739] "Could be +/-infinity or Nan";
```
i would add "Float" somewhere inside the message e.g. "float value could be ...". If i got this right, this version also warns as soon as the result (or whenever norm is called elsewhere) of the operation is top. Maybe we should incorporate this into the message as well " result of operation could be ...".
Apart from that, keep in mind, that in the future we want to support two options on warning with top inputs and warning on top outputs (i guess that is the current version).

Dudeldu commented on src/cdomains/floatDomain.ml:
```diff
@@ -37,18 +37,20 @@ module type FloatDomainBase = sig
+          else v
+        else None
+      | _ -> None
+    in if normed == None then 
```
If the intention is a comparison with top, use the is_top function instead of manually comparing

Dudeldu commented on src/util/options.schema.json:
```diff
@@ -1697,6 +1697,12 @@
+          "title": "warn.float",
+          "description": "float warnings",
+          "type": "boolean",
+          "default": true
```
i guess our new options should be defaulted to `false`, so the behaviour of old examples does not change

Dudeldu commented on tests/regression/56-floats/03-infinity_or_nan.c:
```diff
@@ -0,0 +1,11 @@
+    double data;
+    fscanf(stdin, "%lf", &data);
+    
+    double result1 = data + 1.0; // WARN: Could be +/-infinity or Nan
```
what about some additional examples inside more elaborate constructs
```c
if ((int)(data - 1.)) // WARN: Could be +/-infinity or Nan
    return 0; // `if` requires an actual body, otherwise it is ignored
while ((int)(data - 2.)); // WARN: Could be +/-infinity or Nan
assert ((int)(data - 3.)); // WARN: Could be +/-infinity or Nan
```

brgr replied:
Do they actually give any more information, besides maybe that it's also warning for minus? I personally prefer it to keep it as simple as needed, and in this case we are not checking for any control flow. But I can also change it if you think it's better :)

Dudeldu commented on src/util/messageCategory.ml:
```diff
@@ -21,12 +21,19 @@ type behavior =
 type integer = Overflow | DivByZero [@@deriving eq, ord, hash]
 
+(* Note that this is named '_message' because float itself is a built-in type *)
+type float_message = 
```
this type isn't used anywhere. Maybe the `category` should be extended to `FloatMessage of float_message` ?

brgr committed:
Clean up float domain etc. with fixes from PR

- Better warning message
- use is_top function instead of checking for None in the domain
- revert unrelated change
- make 'warn.float' option not a default option
- make regression test as minimal as possible

brgr committed:
In floatDomain.ml, rearrange of_const function s.t. it finds the norm function

Dudeldu commented on tests/regression/56-floats/03-infinity_or_nan.c:
```diff
@@ -0,0 +1,11 @@
+// PARAM: --enable ana.float.interval
```
you will probably have to enable the now by default disabled float warnings for this test

FelixKrayer replied:
I agree. after that it seems good to merge

brgr replied:
Great catch, thank you! This is also done now

brgr committed:
Add float warning parameter to 03-infinity_or_nan.c regression test

FelixKrayer reviewed (approved):
⊤
